src/python/pants/util/frozendict.py

```diff
@@ -96,6 +106,20 @@ def __lt__(self, other: Any) -> bool:
         # optimising this, by, for instance, sorting on construction.
         return sorted(self._data.items()) < sorted(other._data.items())
 
+    def __or__(self, other: Any) -> FrozenDict[K, V]:
+        if isinstance(other, FrozenDict):
+            other = other._data
+        elif not isinstance(other, Mapping):
+            return NotImplemented
+        return FrozenDict(self._data | other)
+
+    def __ror__(self, other: Any) -> MutableMapping[K, V]:
```

Comment:
I too would prefer it to return a `FrozenDict`, however, I want (even need) to support `dict(foo="bar") | FrozenDict({"baz": 43})`, and for the most part, this would be OK to turn into a FrozenDict as result. However, when the lvalue is a `dict` with unhashable values, it can't be turned into a `FrozenDict`, which leaves us with the alternative to either a) try to freeze the result, and failing that do b) return a regular dict.

The result of `dict | dict` is a `MutableMapping`, but the inputs to the `|` doesn't have to be mutable. (`Mapping | Mapping => MutableMapping`).

Hence, we don't require a mutable mapping as input, but the result will be mutable.

Reply:
> However, when the lvalue is a dict with unhashable values, it can't be turned into a FrozenDict, which leaves us with the alternative to either a) try to freeze the result, and failing that do b) return a regular dict

Hm, does this failure occur in practice? 

If this doesn't occur in practice, it seems like a "safer" API to have `dict | FrozenDict` try to return a `FrozenDict` and emit an error if the values don't work for that. If this becomes annoying in future, we can always relax it, but it'll be hard to tighten it up.

Reply:
[It does.](https://github.com/pantsbuild/pants/actions/runs/9417444019/job/25942770954?pr=21027#step:11:927)
```
>               yield expanded_address, expanded_fields | group_kwargs
E               TypeError: unsupported operand type(s) for |: 'dict' and 'FrozenDict'
```

And the field values here are _raw_, meaning, as provided in the BUILD file thus may be lists or other unhashable values.

The alternative is, of course, to not use `|` here at all. But it was so clean looking so I'm pleased with how it turned out. The thing to keep in mind is just that: `{} | FrozeDict => dict` while `FrozenDict | {} => FrozenDict`. Hmm.. typed out like that I see it's not great. Perhaps I just should strip the frozen status before merging: 
```python
    yield expanded_address, expanded_fields | dict(group_kwargs)
```

Ok, yes, that's probably a better idea. Thanks for all the thorough reviews and persistence to reach the best end result possible.